### PR TITLE
Abstract row writing for simpler rows and expose user quotas

### PR DESCRIPTION
**Pull request by thomasem**
- Abstract common row writing string array joins and writing
- Add `carina quotas` command to display user's quotas

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -851,8 +873,11 @@ func writeClusterHeader(w *tabwriter.Writer) (err error) {
 		"AutoScale",
 		"Status",
 	}
-	s := strings.Join(headerFields, "\t")
+	return writeRow(w, headerFields)
+}
 
+func writeRow(w *tabwriter.Writer, fields []string) (err error) {
+	s := strings.Join(fields, "\t")
```
Comment (rgbkrk) [inline main.go:880]: Yup, this looks pretty good. I figured you would need to do this with our first new set of rows.

**Comment by rgbkrk** [comment]
Looking real good!

```
$ ./carina quotas
MaxClusters         MaxNodesPerCluster
3                   3
```

**Comment by rgbkrk** [comment]
On to the next round which is making a release! There are [some steps to read through](https://github.com/getcarina/carina#releasing), then I'm happy to have you do that too because shipping is pretty great. Our next constellation name for the release name is "Circinus". If you [flip back through our prior release history, you'll see the pattern](https://github.com/getcarina/carina/releases) (aim for alliterative adjectives).

**Comment by thomasem** [comment]
Awesome. I'd be happy to do the release! Thinking "Cirrose Circinus". Version bump to 1.3.0, I assume?

**Comment by rgbkrk** [comment]
That poor liver!

Version bump and name look great. Make sure to create a remote called `release` that you can push to here with. I [see it in our shell script](https://github.com/getcarina/carina/blob/master/release.sh#L54) yet I left it out of the README. Luckily I put that together to fail early before a release is attempted.

**Comment by thomasem** [comment]
Hmmmm... So, installed `github-release`, and running this command from the repo directory - `./release.sh v1.3.0 "Cirrose Circinus"`.

My workspace looks like so:

```
/root
└── /home
    └── /vagrant
        └── /workspace
            └── /carina
```

My initial `GOPATH` - `/home/vagrant/gvm/pkgsets/go1.5/global`
First time running the release command here failed with what looks like `go install` or the `go get ./...` attempting to install things outside of `GOPATH` (in my repository directory). Next, I set `GOPATH` to that directory (`export GOPATH=$(pwd)`, which has worked for builds in the past.

Running the release command I get:

```
Releasing 'v1.3.0' - Cirrose Circinus: The Carina Command Line Client
rm -f bin/*
rm carina
rm: cannot remove ‘carina’: No such file or directory
make: [clean] Error 1 (ignored)
rm carina-linux
rm: cannot remove ‘carina-linux’: No such file or directory
make: [clean] Error 1 (ignored)
rm ca-certificates.crt
rm: cannot remove ‘ca-certificates.crt’: No such file or directory
make: [clean] Error 1 (ignored)
go get ./...
src/github.com/rackspace/gophercloud/openstack/orchestration/v1/stacks/utils.go:13:2: cannot find package "gopkg.in/yaml.v2" in any of:
    /home/vagrant/.gvm/gos/go1.5/src/gopkg.in/yaml.v2 (from $GOROOT)
    /home/vagrant/workspace/carina/src/gopkg.in/yaml.v2 (from $GOPATH)
make: *** [get-deps] Error 1
```

I'm assuming that's another dependency I need to install. So, I removed old stuff, `rm -rf ./src`, then go get that package, `go get gopkg.in/yaml.v2`. After running the release command again I get:

```
Releasing 'v1.3.0' - Cirrose Circinus: The Carina Command Line Client
rm -f bin/*
rm carina
rm: cannot remove ‘carina’: No such file or directory
make: [clean] Error 1 (ignored)
rm carina-linux
rm: cannot remove ‘carina-linux’: No such file or directory
make: [clean] Error 1 (ignored)
rm ca-certificates.crt
rm: cannot remove ‘ca-certificates.crt’: No such file or directory
make: [clean] Error 1 (ignored)
go get ./...
go install: no install location for directory /home/vagrant/workspace/carina outside GOPATH
    For more details see: go help gopath
go install: no install location for directory /home/vagrant/workspace/carina/version outside GOPATH
    For more details see: go help gopath
make: *** [get-deps] Error 1
```

Feels like I'm going in circles... mind sharing any thoughts?

**Comment by thomasem** [comment]
Does this assume I've done `go get github.com/getcarina/carina`? 

from https://github.com/getcarina/carina/blob/master/Makefile#L6:

```
REPO_PATH = ${GOPATH}/src/github.com/${GITHUB_ORG}/${GITHUB_REPO}
```

**Comment by rgbkrk** [comment]
I'm assuming you've done a go get on both libcarina and carina?

Can you run

```
go get ./...
```

within your `github.com/getcarina/carina` directory?

**Comment by thomasem** [comment]
No, I hadn't. That's what I was curious about. Seemed odd to me that I'd `go get github.com/getcarina/carina` from my `carina` repository.

**Comment by rgbkrk** [comment]
Weird. `make get-deps` should be doing this for you.

**Comment by thomasem** [comment]
```
$ make get-deps
go get ./...
src/github.com/rackspace/gophercloud/openstack/orchestration/v1/stacks/utils.go:13:2: cannot find package "gopkg.in/yaml.v2" in any of:
    /home/vagrant/.gvm/gos/go1.5/src/gopkg.in/yaml.v2 (from $GOROOT)
    /home/vagrant/workspace/carina/src/gopkg.in/yaml.v2 (from $GOPATH)
make: *** [get-deps] Error 1
```

**Comment by rgbkrk** [comment]
How's it go if you do that one dep directly?

```
go get gopkg.in/yaml.v2
```

This is pretty odd. Might be worth navigating in to there too.

**Comment by thomasem** [comment]
```
(sn) 09:14:59(-0500) ~/workspace/carina {master} $ rm -rf src/ pkg/
(sn) 09:15:04(-0500) ~/workspace/carina (master) $ go get gopkg.in/yaml.v2
(sn) 09:15:14(-0500) ~/workspace/carina {master} $ make get-deps
go get ./...
go install: no install location for directory /home/vagrant/workspace/carina outside GOPATH
    For more details see: go help gopath
go install: no install location for directory /home/vagrant/workspace/carina/version outside GOPATH
    For more details see: go help gopath
make: *** [get-deps] Error 1
```

**Comment by thomasem** [comment]
```
(sn) 09:15:51(-0500) ~/workspace/carina {master} $ echo $GOPATH
/home/vagrant/workspace/carina
```

**Comment by thomasem** [comment]
How do you release from a clean repository?

**Comment by thomasem** [comment]
Clean meaning fresh box. Only have `go1.5` installed, and just pulled down the repository's master branch, and that's it.

**Comment by rgbkrk** [comment]
Where is the source tree for this repo? It looks like you're running this all from your $GOPATH, when you'll want to be at `/home/vagrant/workspace/carina/src/github.com/getcarina/carina`.

**Comment by thomasem** [comment]
Oh, interesting. So therein lies my disconnect. I guess to develop on `carina` a Golang developer would `go get github.com/getcarina/carina` and then work from the resulting `src/github.com/getcarina/carina`. Whereas, I have `git clone`'ed and worked that way.

**Comment by rgbkrk** [comment]
Yeah, sorry. I should have explained the go flow.

**Comment by thomasem** [comment]
Not a problem, man. Just one of those things someone new to Golang may not realize at first, it's probably somewhere among the tours or tutorials, just hadn't gotten that far yet.

**Comment by thomasem** [comment]
Alright, so here's what I did:

```
(sn) 09:23:51(-0500) ~/workspace $ rm -rf carina
(sn) 09:23:54(-0500) ~/workspace $ mkdir carina
(sn) 09:23:58(-0500) ~/workspace $ cd carina
(sn) 09:24:01(-0500) ~/workspace/carina $ export GOPATH=$(pwd)
(sn) 09:24:08(-0500) ~/workspace/carina $ go get github.com/getcarina/carina
(sn) 09:24:56(-0500) ~/workspace/carina $ ls
bin  pkg  src
(sn) 09:25:00(-0500) ~/workspace/carina $ cd src/
(sn) 09:25:02(-0500) ~/workspace/carina/src $ ls
github.com  gopkg.in
(sn) 09:25:03(-0500) ~/workspace/carina/src $ cd github.com/
alecthomas/ getcarina/  mitchellh/  rackspace/  
(sn) 09:25:03(-0500) ~/workspace/carina/src $ cd github.com/
alecthomas/ getcarina/  mitchellh/  rackspace/  
(sn) 09:25:03(-0500) ~/workspace/carina/src $ cd github.com/
alecthomas/ getcarina/  mitchellh/  rackspace/  
(sn) 09:25:03(-0500) ~/workspace/carina/src $ cd github.com/getcarina/
carina/    libcarina/ 
(sn) 09:25:03(-0500) ~/workspace/carina/src $ cd github.com/getcarina/carina/
(sn) 09:25:48(-0500) ~/workspace/carina/src/github.com/getcarina/carina (master) $ git remote add release git@github.com:getcarina/carina.git
(sn) 09:25:58(-0500) ~/workspace/carina/src/github.com/getcarina/carina (master) $ git fetch --all
Fetching origin
Fetching release
From github.com:getcarina/carina
 * [new branch]      master     -> release/master
(sn) 09:26:07(-0500) ~/workspace/carina/src/github.com/getcarina/carina (master) $ ./release.sh v1.3.0 "Cirrose Circinus"
Releasing 'v1.3.0' - Cirrose Circinus: The Carina Command Line Client
rm -f bin/*
rm carina
rm: cannot remove ‘carina’: No such file or directory
make: [clean] Error 1 (ignored)
rm carina-linux
rm: cannot remove ‘carina-linux’: No such file or directory
make: [clean] Error 1 (ignored)
rm ca-certificates.crt
rm: cannot remove ‘ca-certificates.crt’: No such file or directory
make: [clean] Error 1 (ignored)
go get ./...
CGO_ENABLED=0 go build -a -tags netgo -ldflags '-w -X github.com/getcarina/carina/version.Commit=b611b9dad3f33e1ee9c810ece98a03643b3a4902 -X github.com/getcarina/carina/version.Version=v1.2.1-3-gb611b9d' -o carina .
go get ./...
CGO_ENABLED=0 go build -a -tags netgo -ldflags '-w -X github.com/getcarina/carina/version.Commit=b611b9dad3f33e1ee9c810ece98a03643b3a4902 -X github.com/getcarina/carina/version.Version=v1.2.1-3-gb611b9d' -o carina .
go test -v
=== RUN   TestLoadCache
--- PASS: TestLoadCache (0.00s)
PASS
ok      github.com/getcarina/carina 0.002s
eval ""
./carina --version
Version: v1.2.1-3-gb611b9d
Commit:  b611b9dad3f33e1ee9c810ece98a03643b3a4902
Out with the old, in with the new
Version: v1.2.1-3-gb611b9d
Commit:  b611b9dad3f33e1ee9c810ece98a03643b3a4902
---------------------------------
error: empty token
```

Much closer!

**Comment by rgbkrk** [comment]
Sweet! You... may have to delete the tag if it just created one, or follow along with the rest of the script after we do this.

If it says empty token, it seems it's missing [a GitHub OAuth Token](https://help.github.com/articles/creating-an-access-token-for-command-line-use) which you'll want to set as:

```
export GITHUB_TOKEN=
```

**Comment by thomasem** [comment]
There we go:

```
$ ./release.sh v1.3.0 "Cirrose Circinus"
Releasing 'v1.3.0' - Cirrose Circinus: The Carina Command Line Client
rm -f bin/*
rm carina
rm carina-linux
rm: cannot remove ‘carina-linux’: No such file or directory
make: [clean] Error 1 (ignored)
rm ca-certificates.crt
rm: cannot remove ‘ca-certificates.crt’: No such file or directory
make: [clean] Error 1 (ignored)
go get ./...
CGO_ENABLED=0 go build -a -tags netgo -ldflags '-w -X github.com/getcarina/carina/version.Commit=b611b9dad3f33e1ee9c810ece98a03643b3a4902 -X github.com/getcarina/carina/version.Version=v1.2.1-3-gb611b9d' -o carina .
go get ./...
CGO_ENABLED=0 go build -a -tags netgo -ldflags '-w -X github.com/getcarina/carina/version.Commit=b611b9dad3f33e1ee9c810ece98a03643b3a4902 -X github.com/getcarina/carina/version.Version=v1.2.1-3-gb611b9d' -o carina .
go test -v
=== RUN   TestLoadCache
--- PASS: TestLoadCache (0.00s)
PASS
ok      github.com/getcarina/carina 0.003s
eval ""
./carina --version
Version: v1.2.1-3-gb611b9d
Commit:  b611b9dad3f33e1ee9c810ece98a03643b3a4902
Out with the old, in with the new
Version: v1.2.1-3-gb611b9d
Commit:  b611b9dad3f33e1ee9c810ece98a03643b3a4902
---------------------------------
From github.com:getcarina/carina
 * [new tag]         v1.3.0     -> v1.3.0
```

**Comment by thomasem** [comment]
I went ahead and edited the release and update the links: https://github.com/getcarina/carina/releases/tag/v1.3.0. I noticed the updated links are broken. Will I need to add the binary to the download site, or is that done automatically and just taking a little time?

**Comment by thomasem** [comment]
Welp, just answered my own question. Not broken anymore.

**Comment by thomasem** [comment]
```
$ carina quotas
MaxClusters         MaxNodesPerCluster
20                  3
```

FIN

**Comment by rgbkrk** [comment]
Yep! I'll go ahead and do the homebrew release now as well.

**Comment by rgbkrk** [comment]
By yep I mean to say that the binaries get updated automatically by Travis on a tagged release. Good work!

**Comment by thomasem** [comment]
Thanks! Yeah, that's good, because I don't use OS X (anymore). :stuck_out_tongue:

**Comment by carolynvs** [comment]
Just adding for posterity that pushing an updated chocolatey package is also automatic when you cut a release.

**Comment by thomasem** [comment]
Awesome. I was wondering about that @carolynvs. Thanks!!

**Pull request merged**